### PR TITLE
Drone refresh 2

### PR DESCRIPTION
**Pull request by jbox144**
## About The Pull Request

Reintroduces drones onto the station. Drones now have access to a built in radio, which can recieve from the common and engineering channels, but cannot transmit. Drones can also speak in binary, and understand common.

Drones also now also have a modified lawset, which lets them interact with silicons. Basically, it makes drones less boring.

This may be unbalanced, but I'd like to see this change tested first. I have more changes to drones underway, to better control drones and prevent them being exploited.

## Why It's Good For The Game

Drones are an underappreciated mob, with a lot of potential. I just think they lack any interactability, which often makes drone gameplay very dull.

## Changelog
:cl:
add: Drones get a built-in radio, connected to the common and engineering channels. It cannot transmit however.
add: Drones can be built at mech fabricators, once basic robotics is researched, at the cost of iron, glass and some gold.
tweak: Drones can speak in silicon, and can understand common. Derelict drones cannot speak silicon.
tweak: Drones can speak on binary with the :b prefix. The drone exclusive channel has been moved to the :d prefix. Derelict drones can only use dronechat.
/:cl:

### CODE REVIEW

**Comment by amidoingitright** [comment]
A use for the drone pinger for PDA cartridges at last

**Comment by TheFakeElon** [comment]
Don’t let them speak with IPCs

File: austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+								/datum/language/common = list(LANGUAGE_ATOM))
+	spoken_languages = list(/datum/language/drone = list(LANGUAGE_ATOM),
+							/datum/language/machine = list(LANGUAGE_ATOM))
+	blocked_languages = list()
+
+// Derelict drones don't speak machine, as its too modern for them.
+/datum/language_holder/drone/derelict
+	understood_languages = list(/datum/language/drone = list(LANGUAGE_ATOM),
+								/datum/language/machine = list(LANGUAGE_ATOM),
+								/datum/language/common = list(LANGUAGE_ATOM))
```
Comment (MCterra10) [inline austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm:26]: remove common and machine understanding, please

File: austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	spoken_languages = list(/datum/language/drone = list(LANGUAGE_ATOM))
+
+// Overwrites but references the upstream handle_message proc for binary. This allows drones to speak on binary, but
+/datum/saymode/binary/handle_message(mob/living/user, message, datum/language/language)
+	if(istype(user, /mob/living/simple_animal/drone/derelict)) // Derelict drones can only use drone chat (as per upstream)
+		. = ..()
+		return .
```
Comment (MCterra10) [inline austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm:33]:
```suggestion
		return ..()
```

File: austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+/datum/saymode/drone // Handles the drone exclusive chat
+	key = "d"
+	mode = "drone"
+
+/datum/saymode/drone/handle_message(mob/living/user, message, datum/language/language)
+	if(isdrone(user))
+		var/mob/living/simple_animal/drone/D = user
+		D.drone_chat(message)
+		return FALSE
+	return FALSE // If your not a drone, no drone chat for you
```
Comment (MCterra10) [inline austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm:48]: please add a trailing newline to the end of your file

**Review by MCterra10 — CHANGES_REQUESTED** [review]

**Comment by MCterra10** [comment]
Also this really doesn't fix the problem that drones are just better, more versatile borgs, now in far more literal sense since they can speak on binary and communicate with borgs and the AI. Idk chief.

**Comment by TheFakeElon** [comment]
I’m torn on this one, I agree with Terra here. Silicons are (usually) more fun to play with whereas drones are just the stronger, zero crew interaction version.

**Comment by ZombiesVsPlants** [comment]
ah that's fine, just rebalance borgs to be stronger than humans like intended so you have to trade your strength away for hands.

**Comment by TheFakeElon** [comment]
**”like intended”**

Borgs are supposed to be specialised in certain tasks rather than being a jack of all trades like humans

**Comment by TheFakeElon** [comment]
Drones are an unbalanced mix of the two

**Comment by ZombiesVsPlants** [comment]
standard borg is literally jack of all trades

**Comment by TheFakeElon** [comment]
Standard borg is the jack of no trades. It can’t do anything particularly well.

**Comment by ZombiesVsPlants** [comment]
well then, needs a buff obviously.

**Comment by TheFakeElon** [comment]
Default borg shouldn’t be a jack of all trades. A “general purpose borg” defeats the whole purpose of a borg. I don’t think they should be removed (I’ll miss meme borg) like what TG did but they don’t need to be “buffed” either

**Comment by TheFakeElon** [comment]
ANYWAY this is getting off topic.

**Comment by jbox144** [comment]
I'd be happy with a testmerge

**Comment by MCterra10** [comment]
I'm not *happy* about these changes, but I guess elon can go ahead with a testmerge

**Comment by TheFakeElon** [comment]
I was waiting for your opinion. I was pretty against this but on the other hand, I remember how fun ghost drones were on goon, which were basically this.

**Comment by MCterra10** [comment]
testmerged

**Comment by MCterra10** [comment]
I think it's been testmerged long enough, I'll check with players

**Review by TheFakeElon — APPROVED** [review]

File: austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	spoken_languages = list(/datum/language/drone = list(LANGUAGE_ATOM))
+
+// Overwrites but references the upstream handle_message proc for binary. This allows drones to speak on binary, but
+/datum/saymode/binary/handle_message(mob/living/user, message, datum/language/language)
+	if(istype(user, /mob/living/simple_animal/drone/derelict)) // Derelict drones can only use drone chat (as per upstream)
+		return ..()
+	if(isdrone(user) && user.binarycheck()) // All other drones that use binary
+		user.robot_talk(message)
+		return FALSE
+	return ..() // Anything that isn't a drone is handled normally.
```
Comment (MCterra10) [inline austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm:34]: remove binary chat, this is pretty bad. other than that, I'm okay to merge

**Review by MCterra10 — CHANGES_REQUESTED** [review]

File: austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+	return 0
+
+// Drones understand common and can speak machine, such as with IPCs and borgs.
+/datum/language_holder/drone
+	understood_languages = list(/datum/language/drone = list(LANGUAGE_ATOM),
+								/datum/language/machine = list(LANGUAGE_ATOM),
+								/datum/language/common = list(LANGUAGE_ATOM))
+	spoken_languages = list(/datum/language/drone = list(LANGUAGE_ATOM),
+							/datum/language/machine = list(LANGUAGE_ATOM))
```
Comment (MCterra10) [inline austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm:19]: remove ability to speak machine and this is ok to merge

**Review by MCterra10 — CHANGES_REQUESTED** [review]

File: austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm (hunk 5 above)
Comment (TheFakeElon) [inline austation/code/modules/mob/living/simple_animal/friendly/drone/say.dm:19]:
```suggestion
	spoken_languages = list(/datum/language/drone = list(LANGUAGE_ATOM))
```

**Review by TheFakeElon — APPROVED** [review]

**Comment by TheFakeElon** [comment]
I didn't notice any issues during test merge

**Pull request merged**